wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceRuntimes.java

```diff
@@ -843,16 +843,20 @@ private String sessionUserIdOr(String nameIfNoUser) {
     public void run() {
       long startTime = System.currentTimeMillis();
       LOG.info("Recovering of runtimes is started.");
-
       for (RuntimeIdentity identity : identities) {
         try {
-          recoverOne(infrastructure, identity);
+          // Recover runtime in multi-thread friendly manner
+          // 1. acquire write lock
+          // 2. check if runtime already restored by other threads.
+          // 3. If needed restore runtime.
+          getInternalRuntime(identity.getWorkspaceId());
```

Comment:
It's a bit confusing to call `get` method to recover something. It's basically using it's side-effect.

In `getInternalRuntime`, do we really need writelock for just getting the runtime? Or is it needed only for restore?? If so, I would rather handle writelock in `recoverone` method and make `getInternalRuntime` simpler with just readlock.

I agree that this solution is simpler and safer one, but it's not the most readable and intuitive.

Reply:
I would personally leave the write lock in the (code path of ) `getInternalRuntime()` as well just to be doubly sure we don't mess up when the runtimes are requested concurrently, say from the UI and some automation over REST API.

I agree the naming is a bit confusing here though.

Reply:
I would personally leave the write lock in the (code path of) `getInternalRuntime()` as well just to be doubly sure we don't mess up when the runtimes are requested concurrently, say from the UI and some automation over REST API.

I agree the naming is a bit confusing here though.

Reply:
>In getInternalRuntime, do we really need writelock for just getting the runtime? Or is it needed only for restore??

I assume Yes and No. We need readlock at the beginning and write lock if we need to recover runtime.
Also will need some `double-check locking or so` to ensure a correct response. And all of that doesn't have a lot of 
sense because major lock implementation implements read lock as write lock.


>If so, I would rather handle writelock in recoverone method and make getInternalRuntime simpler with just readlock.

Not sure that is enough. In this case, we also need some checking logic for the method that invokes it.